CPX_Compass/code.py

```diff
@@ -0,0 +1,146 @@
+import board
+import busio
+import adafruit_lsm303
+import neopixel
+import digitalio
+
+BLACK = 0x000000
+RED = 0xFF0000
+GREEN = 0x00FF00
+BLUE = 0x0000FF
+
+i2c = busio.I2C(board.SCL, board.SDA)
+compass = adafruit_lsm303.LSM303(i2c)
+compass.mag_rate = adafruit_lsm303.MAGRATE_30
+#compass.mag_gain = adafruit_lsm303.MAGGAIN_8_1
+
+calibrate_button = digitalio.DigitalInOut(board.BUTTON_A)
+calibrate_button.direction = digitalio.Direction.INPUT
+calibrate_button.pull = digitalio.Pull.UP
+
+
+#-------------------------------------------------------------------------------
+# Replace these two lines with the results of calibration
+#-------------------------------------------------------------------------------
+raw_mins = [1000.0, 1000.0]
+raw_maxes = [-1000.0, -1000.0]
+#-------------------------------------------------------------------------------
+
+
+mins = [0.0, 0.0]
+maxes = [0.0, 0.0]
+corrections = [0.0, 0.0]
+
+pixels = neopixel.NeoPixel(board.NEOPIXEL, 10, brightness=.2, auto_write=False)
+
+led_patterns =  [[4, 5], [5], [6], [7], [8], [9], [9, 0], [0], [1], [2], [3], [4]]
+
+
+def fill(colour):
+    pixels.fill(colour)
+    pixels.show()
+
+
+def warm_up():
+    fill(BLUE)
+    for _ in range(100):
+        _, _, _ = compass.magnetic
+        time.sleep(0.010)
+
+
+def calibrate(do_the_readings):
+    values = [0.0, 0.0]
+    start_time = time.monotonic()
+
+    fill(GREEN)
+
+    # Update the high and low extremes
+    if do_the_readings:
+        while time.monotonic() - start_time < 10.0:
+            values[0], values[1], _ = compass.magnetic
+            values[1] *= -1                           # accel is upside down, so y is reversed
+            if values[0] != 0.0 and values[1] != 0.0: # ignore the random 0 values
+                for i in range(2):
+                    if values[i] < raw_mins[i]:
+                        raw_mins[i] = values[i]
+                    if values[i] > raw_maxes[i]:
+                        raw_maxes[i] = values[i]
+#            time.sleep(0.005)
+
+    # Recompute the correction and the correct mins/maxes
+    for i in range(2):
+        corrections[i] = (raw_maxes[i] + raw_mins[i]) / 2
+        mins[i] = raw_mins[i] - corrections[i]
+        maxes[i] = raw_maxes[i] - corrections[i]
+
+    fill(BLACK)
+
+
+def normalize(value, in_min, in_max):
+    mapped = (value - in_min) * 200 / (in_max - in_min) + -100
+    return max(min(mapped, 100), -100)
+
+
+
+# Setup
+
+warm_up()
+
+if not calibrate_button.value or (raw_mins[0] == 1000.0 and raw_mins[1] == 1000.0):
+    print("Compass calibration")
+    raw_mins[0] = 1000.0
+    raw_mins[1] = 1000.0
+    raw_maxes[0] = -1000.0
+    raw_maxes[1] = -1000.0
+    calibrate(True)
+
+
+    print("Calibration results")
+    print("Update the corresponding lines near the top of the code\n")
+    print("raw_mins = [{0}, {1}]".format(raw_mins[0], raw_mins[1]))
+    print("raw_maxes = [{0}, {1}]".format(raw_maxes[0], raw_maxes[1]))
+    while True:
+        pass
```

Comment:
Why busy wait here?

Reply:
Dave, check please

Reply:
It’s the calibration mode: Calibrate, output the lines to update the code with, and stall. Once the code is updated with the calibration results this code isn’t run and the compass works as a compass. Starting with button A pressed to  drop into this mode as well to,generate new calibration lines.

Reply:
Alternatively, is there a way to exit to the repl immediately?

Reply:
try `sys.exit()`. That's how python would do it and I think it works in CircuitPython too.